staging/src/k8s.io/dynamic-resource-allocation/structured/pools.go

```diff
@@ -27,42 +27,67 @@ import (
 	draapi "k8s.io/dynamic-resource-allocation/api"
 )
 
+func NodeMatches(node *v1.Node, nodeNameToMatch string, allNodesMatch bool, nodeSelector *v1.NodeSelector) (bool, error) {
+	switch {
+	case nodeNameToMatch != "":
+		return node != nil && node.Name == nodeNameToMatch, nil
+	case allNodesMatch:
+		return true, nil
+	case nodeSelector != nil:
+		selector, err := nodeaffinity.NewNodeSelector(nodeSelector)
+		if err != nil {
+			return false, fmt.Errorf("failed to parse node selector %s: %w", nodeSelector.String(), err)
+		}
+		return selector.Match(node), nil
+	}
+
+	return false, nil
+}
+
 // GatherPools collects information about all resource pools which provide
 // devices that are accessible from the given node.
 //
 // Out-dated slices are silently ignored. Pools may be incomplete (not all
 // required slices available) or invalid (for example, device names not unique).
 // Both is recorded in the result.
-func GatherPools(ctx context.Context, slices []*resourceapi.ResourceSlice, node *v1.Node) ([]*Pool, error) {
+func GatherPools(ctx context.Context, slices []*resourceapi.ResourceSlice, node *v1.Node, partitionableDevicesEnabled bool) ([]*Pool, error) {
 	pools := make(map[PoolID]*Pool)
-	nodeName := ""
-	if node != nil {
-		nodeName = node.Name
-	}
 
 	for _, slice := range slices {
+		if !partitionableDevicesEnabled && (len(slice.Spec.CapacityPools) > 0 || slice.Spec.PerDeviceNodeSelection) {
```

Comment:
So, this if statement means: "the feature was enabled, some fields were set, and then the feature was disabled again". We should add a comment to make that understandable. Can this actually happen? Wouldn't the fields have been blanked out by the strategy?

Also, the `continue` means we ignore these. I think that's OK, but should we instead consider them invalid, since that's something we can do here?

I am OK either way.

Reply:
I agree that should not be expected, but we leave the fields in etcd once set with/without feature gate configuration. I have changed to return error in this case. :)